File: sdk/java/src/main/java/io/velocity/examples/SimpleWorker.java
package io.velocity.examples;

import io.velocity.sdk.VelocityWorker;
import io.velocity.sdk.VelocityClient;
import io.velocity.sdk.DurableWorkflow;
import io.velocity.sdk.WorkflowMethod;
import io.velocity.sdk.SignalMethod;

/**
 * Example: Simple task worker using the VELOCITY-WorkFlow Java SDK.
 *
 * Demonstrates:
 *   - Worker registration with options
 *   - Task polling from a task queue
 *   - Task execution with registered handlers
 *   - Error handling
 *   - Shutdown hook for graceful termination
 *
 * Prerequisites:
 *   1. Start the VELOCITY-WorkFlow server:
 *      cd VELOCITY-WorkFlow/src/Velocity.Workflow.Server && dotnet run
 *
 *   2. Build the SDK:
 *      cd VELOCITY-WorkFlow/sdk/java && ./gradlew build
 *
 *   3. Run this example:
 *      ./gradlew run -PmainClass=io.velocity.examples.SimpleWorker
 */
public class SimpleWorker {

    private static final String SERVER_ADDR = "localhost:50051";
    private static final String TASK_QUEUE = "orders";

    // ── Workflow definition ──────────────────────────────────────────────

    @DurableWorkflow(taskQueue = "orders")
    public static class OrderProcessingWorkflow {

        @WorkflowMethod
        public String execute(String input) {
            System.out.printf("[worker] Processing order: %s%n", input);
            // Simulate work
            try { Thread.sleep(50); } catch (InterruptedException ignored) {}
            return "{\"status\": \"shipped\", \"order_id\": " + input + "}";
        }

        @SignalMethod("payment-confirmed")
        public void onPaymentConfirmed(String payload) {
            System.out.printf("[worker] Payment confirmed: %s%n", payload);
        }
    }

    // ── Main entry point ─────────────────────────────────────────────────

    public static void main(String[] args) {
        System.out.println("[worker] Starting VELOCITY-WorkFlow Java worker");
        System.out.printf("[worker] Server: %s | Queue: %s%n", SERVER_ADDR, TASK_QUEUE);

        // Create and configure the worker
        VelocityWorker worker = VelocityWorker.create(SERVER_ADDR, TASK_QUEUE);
        worker.registerWorkflow(OrderProcessingWorkflow.class);

        // Register JVM shutdown hook for graceful termination
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            System.out.println("[worker] Shutdown signal received — stopping worker...");
            worker.stop();
            System.out.println("[worker] Worker shut down cleanly");
        }));

        // Start the worker (begins polling)
        worker.start();

        System.out.printf("[worker] Registered workflows: %s%n", worker.getWorkflowTypes());
        System.out.printf("[worker] Registered signals: %s%n", worker.getSignalHandlers());
        System.out.println("[worker] Polling for tasks... (Ctrl+C to stop)");

        // Poll loop — in production this runs until worker.stop() is called
        try {
            while (worker.isRunning()) {
                try {
                    // Poll for next task (blocking with timeout)
                    // In a full implementation, VelocityWorker.pollAndExecute() handles
                    // task dispatch to registered handlers automatically.
                    Thread.sleep(1000); // Poll interval
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    System.out.println("[worker] Poll loop interrupted");
                    break;
                } catch (Exception e) {
                    System.err.printf("[worker] Error processing task: %s%n", e.getMessage());
                }
            }
        } finally {
            worker.stop();
        }
    }
}

File: sdk/rust/examples/simple_worker.rs
//! Example: Simple task worker using the VELOCITY-WorkFlow Rust SDK.
//!
//! Demonstrates:
//!   - Worker registration with a task queue
//!   - Polling for tasks in a loop
//!   - Executing task logic via registered handlers
//!   - Error handling with typed errors
//!
//! Run:
//!   cd VELOCITY-WorkFlow/sdk/rust
//!   cargo run --example simple_worker

use velocity_sdk::{VelocityClient, WorkflowStatus};
use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

// ── Configuration ────────────────────────────────────────────────────────

const SERVER_ADDR: &str = "localhost:7234";
const TASK_QUEUE: &str = "orders";

// ── Task handler type ────────────────────────────────────────────────────

type TaskHandler = Box<dyn Fn(&Task) -> Result<String, String>>;

struct Task {
    workflow_key: u64,
    workflow_type: String,
    input: Vec<u8>,
}

// ── Task handlers ────────────────────────────────────────────────────────

fn process_order(task: &Task) -> Result<String, String> {
    let input_str = String::from_utf8_lossy(&task.input);
    println!("[worker] Processing order: {}", input_str);
    // Simulate work
    std::thread::sleep(std::time::Duration::from_millis(50));
    Ok(format!(r#"{{"status":"shipped","input":{}}}"#, input_str))
}

fn build_handlers() -> HashMap<&'static str, TaskHandler> {
    let mut handlers: HashMap<&'static str, TaskHandler> = HashMap::new();
    handlers.insert("order-processing", Box::new(process_order));
    handlers
}

// ── Worker loop ──────────────────────────────────────────────────────────

fn main() {
    println!("[worker] Starting VELOCITY-WorkFlow Rust worker");
    println!("[worker] Server: {} | Queue: {}", SERVER_ADDR, TASK_QUEUE);

    let running = Arc::new(AtomicBool::new(true));

    // Set up Ctrl+C handler
    let r = running.clone();
    ctrlc_handler(r);

    let client = VelocityClient::new();
    let handlers = build_handlers();

    println!("[worker] Registered on task queue '{}'", TASK_QUEUE);
    println!("[worker] Polling for tasks... (Ctrl+C to stop)");

    while running.load(Ordering::Relaxed) {
        // Poll for a task from the server
        match poll_task(&client) {
            Ok(Some(task)) => {
                let handler = handlers.get(task.workflow_type.as_str());
                match handler {
                    Some(h) => match h(&task) {
                        Ok(result) => {
                            println!(
                                "[worker] Task '{}' completed successfully",
                                task.workflow_type
                            );
                            let _ = client.complete_workflow(task.workflow_key, result.into_bytes());
                        }
                        Err(e) => {
                            eprintln!("[worker] Task execution error: {}", e);
                        }
                    },
                    None => {
                        eprintln!(
                            "[worker] No handler for task type '{}' — skipping",
                            task.workflow_type
                        );
                    }
                }
            }
            Ok(None) => {
                // No task available — sleep briefly
                std::thread::sleep(std::time::Duration::from_millis(500));
            }
            Err(e) => {
                eprintln!("[worker] Poll error: {}", e);
                std::thread::sleep(std::time::Duration::from_millis(1000));
            }
        }
    }

    client.destroy();
    println!("[worker] Shut down cleanly");
}

fn poll_task(client: &VelocityClient) -> Result<Option<Task>, String> {
    // In a full implementation, this calls client.poll_task(TASK_QUEUE, timeout)
    // For this example, we simulate the poll returning None (no task).
    Ok(None)
}

fn ctrlc_handler(running: Arc<AtomicBool>) {
    // Simplified — in production, use the `ctrlc` crate.
    // This is a placeholder that demonstrates the pattern.
    let _ = running; // Used in real implementation
}
